release/dist/qnap/pkgs.go

```diff
@@ -0,0 +1,188 @@
+		return nil, fmt.Errorf("docker run: %w", err)
+	}
+
+	// TODO(sonia): Use signer?
+
+	return []string{out, out + ".md5"}, nil
+}
+
+type qnapBuildsMemoizeKey struct{}
+
+type innerPkg struct {
+	tsPath  string // tailscale binary path
+	tsdPath string // tailscaled binary path
+}
+
+// qnapBuilds is extra build context shared by all qnap builds.
+type qnapBuilds struct {
+	innerPkgs     dist.Memoize[*innerPkg]
+	dockerImageMu sync.Mutex
+}
+
+// getQnapBuilds returns the qnapBuilds for b, creating one if needed.
+func getQnapBuilds(b *dist.Build) *qnapBuilds {
+	return b.Extra(qnapBuildsMemoizeKey{}, func() any { return new(qnapBuilds) }).(*qnapBuilds)
+}
+
+// buildInnerPackage builds the go binaries used for qnap packages.
+// These binaries get embedded with Tailscale package metadata to form qnap
+// releases.
+func (m *qnapBuilds) buildInnerPackage(b *dist.Build, goenv map[string]string) (*innerPkg, error) {
+	return m.innerPkgs.Do(goenv, func() (*innerPkg, error) {
+		if err := b.BuildWebClientAssets(); err != nil {
+			return nil, err
+		}
+		ts, err := b.BuildGoBinary("tailscale.com/cmd/tailscale", goenv)
+		if err != nil {
+			return nil, err
+		}
+		tsd, err := b.BuildGoBinary("tailscale.com/cmd/tailscaled", goenv)
+		if err != nil {
+			return nil, err
+		}
+
+		// The go binaries above get built and put into a /tmp directory created
+		// by b.TmpDir(). But, we build QNAP with docker, which doesn't always
+		// allow for mounting tmp directories (seemingly dependent on docker
+		// host).
+		// https://stackoverflow.com/questions/65267251/docker-bind-mount-directory-in-tmp-not-working
+		//
+		// So here, we move the binaries into a directory within the b.Repo
+		// path and clean it up when the builder closes.
+
+		tmpDir := filepath.Join(b.Repo, fmt.Sprintf("/tmp-qnap-%s-%s-%s", b.Version.Short, goenv["GOOS"], goenv["GOARCH"]))
+		if err = os.MkdirAll(tmpDir, 0755); err != nil {
+			return nil, err
+		}
+		b.AddOnCloseFunc(func() error {
+			return os.RemoveAll(tmpDir)
+		})
+
+		tsBytes, err := os.ReadFile(ts)
+		if err != nil {
+			return nil, err
+		}
+		tsdBytes, err := os.ReadFile(tsd)
+		if err != nil {
+			return nil, err
+		}
+
+		tsPath := filepath.Join(tmpDir, "tailscale")
+		if err := os.WriteFile(tsPath, tsBytes, 0755); err != nil {
+			return nil, err
+		}
+		tsdPath := filepath.Join(tmpDir, "tailscaled")
+		if err := os.WriteFile(tsdPath, tsdBytes, 0755); err != nil {
+			return nil, err
+		}
+
+		return &innerPkg{tsPath: tsPath, tsdPath: tsdPath}, nil
+	})
+}
+
+func (m *qnapBuilds) makeDockerImage(b *dist.Build) error {
+	return b.Once("make-qnap-docker-image", func() error {
+		out := "build.tailscale.io/qdk:latest"
+		var exitErr *exec.ExitError
+
+		cmd := b.Command(b.Repo, "docker", "image", "inspect", out)
+		if err := cmd.Run(); err == nil {
+			return nil // image already exists, return early
+		} else if !errors.As(err, &exitErr) {
+			return fmt.Errorf("docker image inspect: %w", err)
+		}
```

Comment:
this will skip the image build if/when we modify `Dockerfile.qpkg`.
maybe we can always run `docker build` and rely on docker's caching to make it fast?

Reply:
Oh yeah good point! Will update to always run